sw/device/lib/dif/dif_kmac.h

```diff
@@ -594,6 +594,10 @@ dif_kmac_result_t dif_kmac_mode_kmac_start(
  * If `processed` is not provided then this function will block until the entire
  * message has been processed or an error occurs.
  *
+ * If big-endian mode is enabled for messages (`message_big_endian`) only the
+ * part of the message aligned to 32-bit word boundaries will be byte swapped.
+ * Unaligned leading and trailing bytes will be written into the message as-is.
+ *
```

Comment:
Have we tested this claim? From a hardware perspective, the input data has to be aligned  when written into the block, so I am a bit confused by this comment.

+@eunchan

Reply:
Looks like this is not due to the HW limitation but the way of SW implementing the partial write. In partial write case, SW writes one byte at a time , which leads to the same order even in the byte swapped. Am I right?

Reply:
@eunchan that is correct.

The issue for the DIF API is that, from a software perspective, 'big-endian' mode only has clear semantics when the type of the incoming data is known to the library (e.g. is an array of `uint32_t` for example). When the input is untyped (e.g. `void *`), as it usually is in a general purpose hashing library interface, it isn't clear what effect 'big-endian' mode should have on the incoming data.

One option would be to have typed versions of `absorb`. This would make the effect of 'big-endian' mode much easier to define. For example:

```c
dif_kmac_absorb8(dif_kmac_t *kmac, const uint8_t *msg, size_t len); // untyped alias for absorb8
dif_kmac_absorb8(dif_kmac_t *kmac, const uint8_t *msg, size_t len); // no byte swap
dif_kmac_absorb16(dif_kmac_t *kmac, const uint16_t *msg, size_t len); // byte swap at 2 byte granularity
dif_kmac_absorb32(dif_kmac_t *kmac, const uint32_t *msg, size_t len); // byte swap at 4 byte granularity
dif_kmac_absorb64(dif_kmac_t *kmac, const uint64_t *msg, size_t len); // byte swap at 8 byte granularity
```

That's a bit more API surface though.